kaldb/src/main/java/com/slack/kaldb/metadata/core/KaldbMetadataStore.java

```diff
@@ -61,6 +65,7 @@ public KaldbMetadataStore(
 
     if (shouldCache) {
       cachedModeledFramework = modeledClient.cached();
+      cachedModeledFramework.listenable().addListener(getCacheInitializedListener());
       cachedModeledFramework.start();
```

Comment:
why don't we add `awaitCacheInitialized();` here and wait till the cache has been initialized? vs awaiting before the async ZK operations?

Reply:
Two reasons; first, there exists a race condition where this can result in a deadlock when attempting to initialize multiple copies of a cached store; second, we only need to guarantee a hydrated cache for _some_ of these methods, which results in unnecessary blocking on initialization if this is in the constructor.

Reply:
I see the point in 2) i.e only the get operations need to wait for cache to be initialized. but it shouldn't take very long? Seems like we can miss calls to await if we add new read APIs

But I don't understand the deadlock aspect. Either way it would be good to document it in code?